vertical-pod-autoscaler/enhancements/4016-in-place-updates-support/README.md

```diff
@@ -0,0 +1,109 @@
+- [Proposal](#proposal)
+- [Context](#context)
+- [Design Details](#design-details)
+    - [Applying Updates During Pod Admission](#applying-updates-during-pod-admission)
+    - [Applying Disruption-free Updates](#Applying-disruption-free-updates)
+    - [Applying Disruptive Updates](#applying-disruptive-updates)
+    - [Test Plan](#test-plan)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+
+VPA applies its recommendations with a mutating webhook, during pod creation. It can also evict
+pods expecting that it will apply the recommendation when the pod is recreated. This is a
+disruptive process so VPA has some mechanism to avoid too frequent disruptive updates. 
+
+This proposal allows VPA to apply its recommendations more frequently, with less disruption by
+using in-place update feature which is an alpha feature available in Kubernetes 1.27.
+
+This proposal enables only core uses of  in place updates in VPA with intention of gathering more feedback. Any more
+advanced uses of in place updates in VPA (like applying different recommendations during pod initialization) will be
+introduced as separate enhancement proposals.
+
+### Goals
+
+* Allow VPA to actuate without disruption,
+* Allow VPA to actuate more frequently, when it can actuate without disruption,
+* Allow VPA to actuate in situations where actuation by eviction doesn't work.
+
+### Non-Goals
+
+* Improved handling of injected sidecars
+  * Separate AEP will improve VPAs handling of injected sidecars.
+
+## Proposal
+
+Add new supported values of
+[`UpdateMode`](https://github.com/kubernetes/autoscaler/blob/71b489f5aec3899157b37472cdf36a1de223d011/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L124):
+
+* `InPlaceOnly` and
+* `InPlaceOrRecreate`.
+
+## Context
+
+[In-place update of pod resources KEP](https://github.com/kubernetes/enhancements/issues/1287) is available in alpha in
+[Kubernetes 1.27](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md#api-change-3). The
+feature allows changing container resources while the container is running. It also adds 
+[`ResizePolicy`](https://github.com/kubernetes/api/blob/8360d82aecbc72aa039281a394ebed2eaf0c0ccc/core/v1/types.go#L2448)
+field to Container. This field indicates that the container should be restarted when its resources
+
+## Design Details
+
+VPA Updater will execute in-place updates or eviction-updates in the newly added `InPlaceOnly` and `InPlaceOrRecreate`
+modes. 
+
+### Applying Updates During Pod Admission
+
+For VPAs in `InPlaceOnly` and `InPlaceOrRecreate` modes VPA Admission Controller will apply updates to starting pods,
+like it does for VPAs in `Initial`, `Auto`, and `Recreate` modes.
+
+### Applying Disruption-free Updates
+
+When update changes only resources for which container indicates that it doesn't require a restart then VPA can attempt
+to actuate the change without disrupting the pod. VPA Updater will:
```

Comment:
Do you expect some form of hysteresis here? In some cases a recommendation might flicker between two values and I guess even w/o a full disruption it might not make sense to alter the resources too often (if only to limit the general load in the control plane & kubelet).

Reply:
Yes. VPA will not apply disruption-free update unless the change is by at least 10% of the request. It's possible for that to flicker (90% samples of 100m CPU, 10% samples of 1000 mCPU, percentile changes wildly between those 2 values) but it seems very unlikely.